posts/boost-your-databricks-jobs-and-queries-performance.md

```diff
@@ -0,0 +1,66 @@
+
+One last detail is due to node termination, any data cached will be lost when the cluster completes its downsizing. If cost saving is of a higher priority than perfomance - autoscaling is definetly the way to go! Be sure to read Databricks [best practices about autoscaling](https://docs.databricks.com/clusters/cluster-config-best-practices.html#autoscaling).
+
+# Small cluster with larger VMs is better for complex ETL
+
+A complex ETL job that requires unions and joins across multiple tables will run faster on a cluster that uses very large instance types. Having a reduced number of workers will minimize the amount of data shuffled. Your large workers will also have more RAM available and have less `OUT_OF_MEMORY` errors like below (DLT pipeline):
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-boost-performance-databricks/images/2023/02/databricks-out-of-memory.png)
+
+ A low number of large instance types is always better than a high number of small instance types, for more details, I recommend reading [the Databricks page about cluster size consideration](https://docs.databricks.com/clusters/cluster-config-best-practices.html#cluster-sizing-considerations).
+
+One risk to consider is spot termination of a large worker will have way more of an impact on your cluster's resources than losing a small worker.
+
+# Optimize and Z-ordering are your friends
+
+The [Optimize command](https://docs.databricks.com/delta/optimize.html) is very useful and will consolidate small files into larger ones to improve your query speed. You can even select a subset of your data by inputing a `where` in the `optimize` command. Databricks recommends running `optimize` every night when the EC2 spot prices are low.
+
+Z-ordering can be run on columns that are often used in your query's predicates and that are high cardinality. For example, if you have a table with daily stock price, you will want to z-order on the stock ticker column because there is a high number of tickers. Ticker will often be in your queries as it is a common predicate. Partitionning on the ticker column is not recommended and will create too many small partitions. I recommend reading [the Databricks page](https://docs.databricks.com/delta/data-skipping.html#what-is-z-ordering) about z-ordering for more information.
+
+# Cache your data!
+
+This advice will sound basic but caching your data using disk caching is still the standard method to make your queries faster. Databricks automatically cache your data based on your queries but you can also preload data beforehand by using the `CACHE SELECT` command. It is surely very specific to a query and might not fit your use cases especially if the query you want to cache is built dynamically. But for a specific query that is run often, pre-caching it will have a big impact.
```

Comment:
```suggestion
This advice will sound basic but caching your data using disk caching is still the standard method to make your queries faster. Databricks automatically caches your data based on your queries but you can also preload data beforehand by using the `CACHE SELECT` command. It is very specific to a query and might not fit your use case if the query you want to cache is built dynamically. But for a specific query that is run often, pre-caching it will have a big impact.
```